### PR TITLE
feat(cargo-update): `--precise` to allow yanked versions

### PR DESCRIPTION
**Pull request by weihanglo**
### What does this PR try to resolve?

Address #4225

Supports specifying yanked versions via the `--precise` flag of `cargo update`.
This is still behind `-Zunstable-options` nightly flag.

### How should we test and review this PR?

Review this commit by commit.

As Jacob mentioned in <https://github.com/rust-lang/cargo/issues/4225#issuecomment-1863755272>, only indexes in registry sources can be yanked. None of [git source](https://github.com/rust-lang/cargo/blob/7bb7b539558dc88bea44cee4168b6269bf8177b0/src/cargo/sources/git/source.rs#L410-L412), [path source](https://github.com/rust-lang/cargo/blob/7bb7b539558dc88bea44cee4168b6269bf8177b0/src/cargo/sources/path.rs#L609-L611), [directory source](https://github.com/rust-lang/cargo/blob/7bb7b539558dc88bea44cee4168b6269bf8177b0/src/cargo/sources/directory.rs#L264-L266) can. It should be pretty safe not checking sources.

The other thing needing a review is the warning. It is pretty slim at this moment, but a big warning can be added if we want to scare people away from using yanked versions 👻.

### CODE REVIEW

File: src/cargo/sources/registry/mod.rs [hunk 1]
```diff
@@ -775,24 +779,46 @@ impl<'cfg> Source for RegistrySource<'cfg> {
...
+                    } else if self.yanked_whitelist.contains(&s.package_id()) {
+                        callback(s);
+                    } else if req.is_precise() {
+                        precise_yanked_in_use = true;
+                        if self.config.cli_unstable().unstable_options {
+                            callback(s);
+                        }
                     }
                 }))?;
+            if precise_yanked_in_use {
```
Comment (weihanglo) [inline src/cargo/sources/registry/mod.rs:807]: Oh this is not an ideal place to emit a warning, or we need to track if it is already emitted.

**Comment by Eh2406** [comment]
Leaving open for feedback on how to emit the warning. But r=me when ready.

File: src/cargo/sources/registry/mod.rs (hunk 1 above)
Comment (weihanglo) [reply src/cargo/sources/registry/mod.rs:814]: Thanks for the review, @Eh2406!

Yeah. I guess we got a workaround for deduplicating warnings. See the last two commits.

File: src/cargo/sources/registry/mod.rs [hunk 2]
```diff
@@ -775,24 +786,49 @@ impl<'cfg> Source for RegistrySource<'cfg> {
...
+                    .precise_version()
+                    .expect("--precise <yanked-version> in use");
+                if self.selected_precise_yanked.insert((name, version.clone())) {
+                    let source = self.source_id();
+                    let mut shell = self.config.shell();
+                    shell.warn(format_args!(
+                        "yanked package `{name}@{version}` is selected by the `--precise` flag from {source}",
+                    ))?;
+                    shell.note("it is not recommended to depend on a yanked version")?;
+                    shell.note("if possible, try other SemVer-compatbile versions")?;
```
Comment (Eh2406) [inline src/cargo/sources/registry/mod.rs:829]: `compatbile` -> `compatible`

**Comment by Eh2406** [comment]
Looks good to me, fix the typo and merge.

File: tests/testsuite/update.rs [hunk 3]
```diff
@@ -1400,10 +1400,34 @@ fn precise_yanked() {
...
+    p.cargo("update --precise 0.1.1 bar")
+        .masquerade_as_nightly_cargo(&["--precise <yanked-version>"])
+        .arg("-Zunstable-options")
+        .with_stderr(
+            "\
+[UPDATING] `dummy-registry` index
+[WARNING] yanked package `bar@0.1.1` is selected by the `--precise` flag from registry `dummy-registry`
+[NOTE] it is not recommended to depend on a yanked version
+[NOTE] if possible, try other SemVer-compatbile versions
+[UPDATING] bar v0.1.0 -> v0.1.1
```
Comment (epage) [inline tests/testsuite/update.rs]: nit: this is written from the perspective of the user knowing its yanked.  I think I would write it from the perspective that they don't know
```
[WARNING] Selected package `bar@0.1.1` was yanked by the author
[NOTE] if possible, try a compatible non-yanked version
```

File: src/doc/man/cargo-update.md [hunk 4]
```diff
@@ -42,6 +42,10 @@ Cannot be used with `--precise`.
 When used with _spec_, allows you to specify a specific version number to set
 the package to. If the package comes from a git repository, this can be a git
 revision (such as a SHA hash or tag).
+
+While not recommended, specifying a yanked version of a package is allowed (nightly only).
+When possible, try other non-yanked SemVer-compatible versions, or seek helps
```
Comment (epage) [inline src/doc/man/cargo-update.md:47]: `helps` should be `help`

File: src/doc/man/cargo-update.md [hunk 5]
```diff
@@ -42,6 +42,10 @@ Cannot be used with `--precise`.
 When used with _spec_, allows you to specify a specific version number to set
 the package to. If the package comes from a git repository, this can be a git
 revision (such as a SHA hash or tag).
+
+While not recommended, specifying a yanked version of a package is allowed (nightly only).
```
Comment (epage) [inline src/doc/man/cargo-update.md:46]: nit: I think it reads a little better to say "While not recommended, you can specify a yanked version of a package".

File: src/doc/src/reference/resolver.md [hunk 6]
```diff
@@ -325,9 +325,11 @@ the `links` field if your library is in common use.
 
 [Yanked releases][yank] are those that are marked that they should not be
 used. When the resolver is building the graph, it will ignore all yanked
-releases unless they already exist in the `Cargo.lock` file.
+releases unless they already exist in the `Cargo.lock` file, or are explicitly
```
Comment (epage) [inline src/doc/src/reference/resolver.md:328]: nit: I think the `,` in here can go as well

**Comment by epage** [comment]
(no need to wait on me to approve fixes for any of my comments)

File: tests/testsuite/update.rs (hunk 3 above)
Comment (weihanglo) [reply tests/testsuite/update.rs]: Putting `source` there was to disambiguate when needed, but maybe that's too much. And we can always improve it later on.

**Comment by weihanglo** [comment]
Addressed all the wording suggestions. Thanks!

@bors r=Eh2406

**Comment by ehuss** [comment]
Is #4225 intended to be the tracking issue for this?

**Comment by weihanglo** [comment]
> Is #4225 intended to be the tracking issue for this?

Sort of? I expect a relatively short unstable period then we merge it. I am not sure if everybody is happy about insta-stable.

**Comment by epage** [comment]
I don't see us learning much from this being unstable and am good with having an FCP on this now.

**Comment by epage** [comment]
The only change I can see maybe eventually doing is making yanked versions candidates but making them the lowest priority.  So long as we don't close the door on that which I don't think this change does, I'm good.

**Pull request merged**